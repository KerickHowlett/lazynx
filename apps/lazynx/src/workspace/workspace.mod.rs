mod workspace_store;

mod workspace_view_widget;
pub use workspace_view_widget::WorkspaceViewWidget;

mod workspace_tab_widget;
pub use workspace_tab_widget::WorkspaceTabWidget;

#[cfg(test)]
#[path = "./test_utils/workspace_test_utils.mod.rs"]
pub mod test_bed;
