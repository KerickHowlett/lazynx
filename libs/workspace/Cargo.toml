[package]
name = "workspace"
version = "0.1.0"
edition = "2021"

[dependencies]
# Local Dependencies
app_config = { path = "../../libs/shared/config" }

# Third-Party Dependencies
chrono = "0.4.38"
color-eyre = "0.6.3"
crossterm = { version = "0.28.1", features = ["serde", "event-stream"] }
ratatui = { version = "0.29.0", features = ["serde", "macros"] }

[dev-dependencies]
insta = "1.42.0"
