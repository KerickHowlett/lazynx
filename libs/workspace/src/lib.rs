mod workspace_view_widget;
pub use workspace_view_widget::WorkspaceViewWidget;

mod workspace_tab_widget;
pub use workspace_tab_widget::WorkspaceTabWidget;
