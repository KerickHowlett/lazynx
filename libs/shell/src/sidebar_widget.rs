use std::rc::Rc;

use ratatui::{
    buffer::Buffer,
    prelude::{Constraint, Direction, Layout, Rect},
    widgets::Widget,
};

use workspace::WorkspaceTabWidget;

#[derive(Default, Clone)]
pub struct SidebarWidget {
    workspace_tab: WorkspaceTabWidget,
}

impl SidebarWidget {
    pub fn init(&mut self) -> color_eyre::eyre::Result<()> {
        self.workspace_tab.init()?;
        Ok(())
    }

    fn create_layout(&self, area: Rect) -> Rc<[Rect]> {
        return Layout::default()
            .direction(Direction::Vertical)
            .constraints(vec![Constraint::Length(3), Constraint::Fill(1)])
            .split(area);
    }
}

impl Widget for SidebarWidget {
    fn render(self, area: Rect, buf: &mut Buffer)
    where
        Self: Sized,
    {
        let chunks = self.create_layout(area);
        self.workspace_tab.render(chunks[0], buf);
    }
}

#[cfg(test)]
mod sidebar_widget_tests {
    use super::SidebarWidget;

    use color_eyre::eyre::Result;
    use insta::assert_snapshot;

    use test_utils::WidgetTestBed;
    use workspace::test_bed::WorkspaceTestBed;

    struct TestBed {
        widget: WidgetTestBed<SidebarWidget>,
        workspace: WorkspaceTestBed,
    }

    impl Default for TestBed {
        fn default() -> Self {
            return TestBed {
                workspace: WorkspaceTestBed::default(),
                widget: WidgetTestBed::<SidebarWidget>::new(30, 50),
            };
        }
    }

    #[test]
    fn test_sidebar_widget_render() -> Result<()> {
        let mut test_bed = TestBed::default();
        test_bed.workspace.setup();

        test_bed.widget.widget.init().unwrap();
        test_bed
            .widget
            .terminal
            .draw(|f| f.render_widget(test_bed.widget.widget, f.area()))
            .unwrap();

        assert_snapshot!(test_bed.widget.terminal.backend());

        test_bed.workspace.restore();

        Ok(())
    }
}
