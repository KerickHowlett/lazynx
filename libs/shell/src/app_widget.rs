use ratatui::{
    buffer::Buffer,
    layout::Rect,
    prelude::{Constraint, Direction, Layout},
    widgets::Widget,
};

use color_eyre::eyre::Result;
use workspace::WorkspaceViewWidget;

use crate::sidebar_widget::SidebarWidget;

pub trait IAppWidget {
    fn init(&mut self) -> Result<()>;
}

#[derive(Clone, Default)]
pub struct AppWidget {
    sidebar: SidebarWidget,
    workspace: WorkspaceViewWidget,
}

impl IAppWidget for AppWidget {
    fn init(&mut self) -> color_eyre::eyre::Result<()> {
        self.sidebar.init()?;
        Ok(())
    }
}

impl Widget for AppWidget {
    fn render(self, area: Rect, buf: &mut Buffer)
    where
        Self: Sized,
    {
        let chunks = Layout::default()
            .direction(Direction::Horizontal)
            .constraints(vec![Constraint::Min(38), Constraint::Percentage(75)])
            .split(area);

        self.sidebar.render(chunks[0], buf);
        self.workspace.render(chunks[1], buf);
    }
}

#[cfg(test)]
mod app_widget_tests {
    use super::{AppWidget, IAppWidget};

    use color_eyre::eyre::Result;
    use insta::assert_snapshot;

    use test_utils::{mocks::MOCK_DATE, WidgetTestBed};
    use workspace::{test_bed::WorkspaceTestBed, WorkspaceViewWidget};

    struct TestBed {
        widget: WidgetTestBed<AppWidget>,
        workspace: WorkspaceTestBed,
    }

    impl Default for TestBed {
        fn default() -> Self {
            let mut widget = AppWidget::default();
            widget.workspace = WorkspaceViewWidget::new(*MOCK_DATE);
            let widget_testbed = WidgetTestBed::<AppWidget>::new(100, 50).with_widget(widget);

            return TestBed {
                workspace: WorkspaceTestBed::default(),
                widget: widget_testbed,
            };
        }
    }

    #[test]
    fn test_app_widget_render() -> Result<()> {
        let mut test_bed = TestBed::default();
        test_bed.workspace.setup();

        test_bed.widget.widget.init().unwrap();
        test_bed
            .widget
            .terminal
            .draw(|f| f.render_widget(test_bed.widget.widget, f.area()))
            .unwrap();

        assert_snapshot!(test_bed.widget.terminal.backend());

        test_bed.workspace.restore();

        Ok(())
    }
}
