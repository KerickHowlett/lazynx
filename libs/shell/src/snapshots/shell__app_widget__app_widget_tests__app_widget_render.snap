---
source: libs/shell/src/app_widget.rs
expression: test_bed.widget.terminal.backend()
---
"╭─[1]─Workspace──────────────────────╮╭Workspace───────────────────────────────────────────────────╮"
"│   test                             ││                                                            │"
"╰────────────────────────────────────╯│   _                     _   _                              │"
"                                      │  | |                   | \ | |                             │"
"                                      │  | |     __ _ _____   _|  \| |_  __                        │"
"                                      │  | |    / _` |_  / | | | . ` \ \/ /                        │"
"                                      │  | |___| (_| |/ /| |_| | |\  |>  <                         │"
"                                      │  \_____/\__,_/___|\__, \_| \_/_/\_\                        │"
"                                      │                    __/ |                                   │"
"                                      │                   |___ /                                   │"
"                                      │                                                            │"
"                                      │  Copyright © 2025 Kerick Howlett                           │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      │                                                            │"
"                                      ╰────────────────────────────────────────────────────────────╯"
